pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/SegmentPrunerFactory.java

```diff
@@ -97,4 +105,44 @@ private static PartitionSegmentPruner getPartitionSegmentPruner(TableConfig tabl
       return new PartitionSegmentPruner(tableNameWithType, partitionColumn, propertyStore);
     }
   }
+
+  @Nullable
+  private static TimeRangeSegmentPruner getTimeRangeSegmentPruner(TableConfig tableConfig,
+      ZkHelixPropertyStore<ZNRecord> propertyStore) {
+    String tableNameWithType = tableConfig.getTableName();
+    SegmentsValidationAndRetentionConfig validationConfig = tableConfig.getValidationConfig();
+    if (validationConfig == null) {
+      LOGGER.warn("Cannot enable time range pruning without validation config for table: {}",
+          tableNameWithType);
+      return null;
+    }
+    String timeColumn = validationConfig.getTimeColumnName();
+    if (timeColumn == null) {
+      LOGGER.warn("Cannot enable time range pruning without time column for table: {}",
+          tableNameWithType);
+      return null;
+    }
+
+    LOGGER.info("Using TimeRangePruner on time column: {} for table: {}", timeColumn,
+        tableNameWithType);
+    return new TimeRangeSegmentPruner(tableConfig, propertyStore);
+  }
+
+  private static List<SegmentPruner> sortSegmentPruners(List<SegmentPruner> pruners) {
+    // If there's multiple pruners, move time range pruners to the front。
+    // Partition pruner run time is proportional to input # of segments while time range pruner is not,
+    // Prune based on time range first will have a smaller input size for partition pruners, so have better performance.
+    List<SegmentPruner> sortedPruners = new ArrayList<>();
+    for (SegmentPruner pruner : pruners) {
+      if (pruner instanceof TimeRangeSegmentPruner) {
+        sortedPruners.add(pruner);
+      }
+    }
+    for (SegmentPruner pruner: pruners) {
+      if (!(pruner instanceof TimeRangeSegmentPruner)) {
```

Comment:
This check is not needed again right?

Reply:
The first for loop will add `TimeRangeSegmentPruner` to the sorted result, this loop are adding non-timeRange pruners. To avoid duplicated pruners, I think this is still needed.